test/unit/math/opencl/copy_test.cpp

```diff
@@ -8,6 +8,151 @@
+  volatile int i = 19;
+  volatile double no_opt = a(i);
+  a(i) = no_opt;
+  no_opt = b(i);
+  b(i) = no_opt;
+
+  MatrixXd correct = a + b;
+  MatrixXd result = from_matrix_cl(c_cl);
+  for (int i = 0; i < N * N; i++) {
+    EXPECT_EQ(result(i), correct(i));
+  }
+  no_opt = w.array().sum();
+}
+
+TEST(MathMatrixGPU, copy_destroyed_to_matrix_cl_eval) {
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+  using Eigen::MatrixXd;
+  using stan::math::from_matrix_cl;
+  using stan::math::matrix_cl;
+  int N = 100;
+  MatrixXd a = MatrixXd::Random(N, N);
+  MatrixXd b = MatrixXd::Random(N, N);
+  matrix_cl<double> c_cl
+      = stan::math::to_matrix_cl((a + b).eval());  // the problem
+  // attempt to scramble the memory that was used by the temporary
+  MatrixXd w = a - b;
+
+  // make compiler think a and b were changed, so second sum can not be
+  // optimized away
+  volatile int i = 19;
+  volatile double no_opt = a(i);
+  a(i) = no_opt;
+  no_opt = b(i);
+  b(i) = no_opt;
+
+  MatrixXd correct = a + b;
+  MatrixXd result = from_matrix_cl(c_cl);
+  for (int i = 0; i < N * N; i++) {
+    EXPECT_EQ(result(i), correct(i));
+  }
+  no_opt = w.array().sum();
+}
+
+TEST(MathMatrixGPU, copy_destroyed_to_matrix_cl_scalar) {
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+  using Eigen::MatrixXd;
+  using stan::math::from_matrix_cl;
+  using stan::math::matrix_cl;
+  double a = 1.0;
+  double b = 2.0;
+  double c = 3.0;
+  double d = 4.0;
+  matrix_cl<double> c_cl
+      = stan::math::to_matrix_cl(a + b + c + d);  // the problem
+  // attempt to scramble the memory that was used by the temporary
+  double f = a - b - c - d;
+
+  // make compiler think a and b were changed, so second sum can not be
+  // optimized away
+  volatile double no_opt = a;
+  a = no_opt;
+  no_opt = b;
+  b = no_opt;
+
+  double correct = a + b + c + d;
+  MatrixXd result = from_matrix_cl(c_cl);
+  EXPECT_EQ(result(0), correct);
+  no_opt = f + b + c + d;
+}
+
+TEST(MathMatrixGPU, copy_destroyed_constructor_scalar) {
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+  using Eigen::MatrixXd;
+  using stan::math::from_matrix_cl;
+  using stan::math::matrix_cl;
+  double a = 1.0;
+  double b = 2.0;
+  double c = 3.0;
+  double d = 4.0;
+  matrix_cl<double> c_cl(a + b + c + d);  // the problem
+  // attempt to scramble the memory that was used by the temporary
+  double f = a - b - c - d;
+
+  // make compiler think a and b were changed, so second sum can not be
+  // optimized away
+  volatile double no_opt = a;
+  a = no_opt;
+  no_opt = b;
+  b = no_opt;
+
+  double correct = a + b + c + d;
+  MatrixXd result = from_matrix_cl(c_cl);
+  EXPECT_EQ(result(0), correct);
+  no_opt = f + b + c + d;
+}
+
+TEST(MathMatrixGPU, std_vector_copy_destroyed_constructor) {
+  using Eigen::MatrixXd;
+  using stan::math::from_matrix_cl;
+  using stan::math::matrix_cl;
+
+  matrix_cl<double> c_cl(std::vector<double>({1, 2, 3, 4, 5, 6}), 2,
+                         3);  // the problem
```

Comment:
I think for the sake of clarity you should not shape this vector as 2x3 matrix on GPU. Since you made a constructor with default shape you can probably use it.

Does this reproduce the bug? I think it is unlikely to if you do not try to overwrite the values of the temporary object.

Reply:
Will use the default constructor, Yes, it reproduces the bug, same as above.